### DISCUSSION TITLE
Aspire to become a full fledged IaC product, or focus on integration with IaC products?

### DISCUSSION THREAD

**Discussion by VincentH-Net** [microsoft/aspire#2014] [Ideas]
Should Aspire aim to include full fledged IaC product features, or rather focus on smooth integration with existing IaC products such as Pulumi?

I'm very interested in how the Aspire team & community thinks about this. I had an interesting thought exchange about this with @ellismg (Azure SDKs team) on the `azure-dev` repo; I'd like to share this here as a starting point:

I asked [this question](https://github.com/Azure/azure-dev/issues/93#issuecomment-1904152032):
> It would be great if devs could use .NET Aspire, which enables automated deployment via AZD, together with Pulumi. I'm working on a project where I'd like to do that.
>
>So far I have not seen concrete advice on how to combine / plans for supporting that in the bits - see [discussion on pulumi repo](https://github.com/pulumi/pulumi/discussions/15015)

And Ellis answered with [a great overview of gaps in AZD and ideas about how to address those](https://github.com/Azure/azure-dev/issues/93#issuecomment-1910851844), in part around Preview 4👍

My full response starts [here](https://github.com/Azure/azure-dev/issues/93#issuecomment-1916310346); I liked a lot of the ideas, but also raised this point:
> However while with the AZD support Aspire extends into Azure IaC to a degree, Aspire is not a full fledged IaC solution.
> > (Ellis) my general view is that Aspire is/can be this (= a full fledged IaC solution)
>
> Hmmm... my grok (so far) is that Aspire helps simplify multiple aspects of distributed application development (IaC being one), while Pulumi is fully focused on IaC. Think e.g. about all the different environments a solution exists in with their respective infrastructure variations and partial IaC reuse across them, versus just two infrastructure variations: local and deployed.
>
> As a .NET / Azure distributed application architect I like to have as much power as possible, as soon as possible, with early maturity, to reduce complexity and effort. A clear product focus helps for that; from that perspective it would appeal to me to have Aspire become truly great for development-centric aspects of distributed applications, including the first step / core IaC, and focus on a smooth and flexible integration with specialized infrastructure solutions such as Pulumi, rather than trying to make Aspire a full IaC competitor. As seen on many occasions, combining strengths and clear focus makes an ecosystem more valuable than adding more alternatives that have a bigger overlap with a less intense focus. It results in earlier maturity as well.
>
> Just food for thought... not disagreeing with you here.

So I'm interested to hear the wider Aspire team and community in this and learn / evolve the thinking. I hope that sharing and engaging in this will help us wrap our minds around the technologies, and it may help to us determine a course in solution technology roadmaps.

Thanks in advance for sharing!

**Comment by VincentH-Net** [comment]
FYI @davidfowl [wrote](https://twitter.com/davidfowl/status/1753951030210093120):
> I plan to carve out time to build aspire integration with pulumi

**Comment by rebeccapowell** [comment]
We are heavily invested in Pulumi and one of the use cases it solves is separated permissions and roles for application execution vs deployment and management of Azure resources. As I understand it, these are not separable in Aspire currently?

I'd like to see a Pulumi Aspire bridge and then you get the best of both worlds.

**Reply by davidfowl** [reply]
We're not going to invest in building this bridge, especially not short term. I spent a little bit of time looking at what integration looks like here https://github.com/davidfowl/AspirePulumi. It's not complete by any means but it'll take someone motivated and someone using pulumi to build something a bit more complete. 

If you're interested I would suggest modeling it after what we've done for the azure resources and what the CDK team is doing with the cloud formation and [CDK resource types](https://github.com/dotnet/aspire/pull/2225).

Happy to help anyone that wants to take a stab at this.
